driver-core/src/test/unit/com/mongodb/AuthConnectionStringTest.java

```diff
@@ -93,31 +92,65 @@ private void testValidUris() {
+            assertString("credential.username", credential.getUserName());
+            assertMechanismProperties(credential);
+
+            // Passwords for certain auth mechanisms are ignored.
+            String password = credential.getPassword() != null ? new String(credential.getPassword()) : null;
+            if (password != null) {
+                assertString("credential.password", password);
+            }
 
-        // Passwords for certain auth mechanisms are ignored.
-        String password = credential.getPassword() != null ? new String(credential.getPassword()) : null;
-        if (password != null) {
-            assertString("auth.password", password);
+            assertMechanism("credential.mechanism", credential.getMechanism());
         }
-        if (definition.get("options").isDocument()) {
-            for (Map.Entry<String, BsonValue> option : definition.getDocument("options").entrySet()) {
-                if (option.getKey().equals("authmechanism")) {
-                    String expected = option.getValue().asString().getValue();
-                    if (expected.equals("MONGODB-CR")) {
-                        assertNotNull(connectionString.getCredential());
-                        assertNull(connectionString.getCredential().getAuthenticationMechanism());
+    }
+
+    private void assertString(final String key, final String actual) {
+        BsonValue expected = getExpectedValue(key);
+
+        if (expected.isNull()) {
+            assertNull(String.format("%s should be null", key), actual);
+        } else if (expected.isString()) {
+            String expectedString = expected.asString().getValue();
+            assertTrue(String.format("%s should be %s but was %s", key, actual, expectedString), actual.equals(expectedString));
+        } else {
+            assertTrue(String.format("%s should be %s but was %s", key, actual, expected), false);
+        }
+    }
+
+    private void assertMechanism(final String key, final String actual) {
+        BsonValue expected = getExpectedValue(key);
+
+        if (expected.isString() && expected.asString().getValue().equals("MONGODB-CR")) {
+            assertNull(String.format("%s should be null when the expected mechanism is MONGODB-CR", key), actual);
+        } else {
+            assertString(key, actual);
+        }
+    }
+
+    private void assertMechanismProperties(final MongoCredential credential) {
+        BsonValue expected = getExpectedValue("credential.mechanism_properties");
+
+        if (!expected.isNull()) {
+            BsonDocument document = expected.asDocument();
+            for (String key : document.keySet()) {
+                if (document.get(key).isString()) {
+                    String expectedValue = document.getString(key).getValue();
+                    if (credential.getMechanism().equals("GSSAPI") && key.equals("SERVICE_NAME") && expectedValue.equals("mongodb")) {
```

Comment:
This deserves a comment explaining why we need this condition.

Reply:
I have added a comment.

Reply:
Add "The default value is applied in the authenticator, not in MongoCredential"  to make it crystal clear.